functions/go/set-namespace/transformer/namespace.go

```diff
@@ -18,206 +18,310 @@ import (
+		return true, nil
 	}
 	// Update "namespace" to the proper resources.
-	oldNamespaces := map[string]struct{}{}
-	count := tc.Transform(rl.Items, oldNamespaces)
-	result := tc.LogResults(rl, count, oldNamespaces)
-	rl.Results = append(rl.Results, result)
+	results := tc.Transform(rl.Items)
+	rl.Results = append(rl.Results, results...)
 	return true, nil
 }
 
-func (p *NamespaceTransformer) LogResults(rl *fn.ResourceList, count int, oldNamespaces map[string]struct{}) *fn.Result {
-	if len(p.Errors) != 0 {
-		errMsg := strings.Join(p.Errors, "\n")
-		return fn.GeneralResult(errMsg, fn.Error)
-	}
-	if dupErr := fn.CheckResourceDuplication(rl); dupErr != nil {
-		return fn.ErrorResult(dupErr)
-	}
-	if count == 0 {
-		msg := fmt.Sprintf("all namespaces are already %q. no value changed", p.NewNamespace)
-		return fn.GeneralResult(msg, fn.Info)
-	}
-
-	oldNss := []string{}
-	for oldNs := range oldNamespaces {
-		oldNss = append(oldNss, `"`+oldNs+`"`)
-	}
-	msg := fmt.Sprintf("namespace %v updated to %q, %d value(s) changed", strings.Join(oldNss, ","), p.NewNamespace, count)
-	return fn.GeneralResult(msg, fn.Info)
+// SetNamespace defines structs to parse KRM resource "SetNamespace" (the custom function config) and "ConfigMap" data.
+// it provides the method "Config" to read the function configs from ResourceList.FunctionConfig
+// it provides the method "Transform" to change the "namespace" and update the "config.kubernetes.io/depends-on" annotation.
+type SetNamespace struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	NewNamespace      string            `json:"namespace,omitempty"`
+	NamespaceMatcher  string            `json:"namespaceMatcher,omitempty"`
+	Data              map[string]string `json:"data,omitempty"`
 }
 
-// Config gets the attributes from different FunctionConfig formats.
-func (p *NamespaceTransformer) Config(o *fn.KubeObject) error {
+// Config gets the new namespace from FunctionConfig. It accepts three types of FunctionConfig:
+// 1. A ConfigMap object's .data.namespace
+// 2. A ConfigMap named "kptfile.kpt.dev" object's .data.name
+// 3. A SetNamespace object's .namespace
+func (p *SetNamespace) Config(o *fn.KubeObject) error {
 	switch {
 	case o.IsEmpty():
-		return fmt.Errorf("FunctionConfig is missing. Expect `ConfigMap` or `SetNamespace`")
-	case o.IsGVK("v1", "ConfigMap"):
-		p.NewNamespace = o.NestedStringOrDie("data", "namespace")
-		if p.NewNamespace == "" {
-			if o.GetName() == "kptfile.kpt.dev" {
-				p.NewNamespace = o.NestedStringOrDie("data", "name")
-				if p.NewNamespace == "" {
-					return fmt.Errorf("`data.name` should not be empty")
-				}
-			} else {
-				return fmt.Errorf("`data.namespace` should not be empty")
-			}
+		return fmt.Errorf("FunctionConfig is missing. Expect `ConfigMap(v1)` or `%s(%s/%s)`",
+			fnConfigKind, fnConfigGroup, fnConfigVersion)
+	case o.IsGVK("", "v1", "ConfigMap"):
+		o.AsOrDie(&p)
```

Comment:
If we take my suggest above to separate `SetNamespace` from `ConfigMap` structs, we can initialize the `ConfigMap` type here and then copy its data field into `p`'s fields. 

We're already copying the `data` field into `p`'s other fields, so the separation wouldn't add any complexity.

Reply:
done.